api/src/main/java/io/druid/segment/loading/DataSegmentPusher.java

```diff
@@ -89,13 +102,21 @@ default String makeIndexPathName(DataSegment dataSegment, String indexName)
   // If above format is ever changed, make sure to change it appropriately in other places
   // e.g. HDFSDataSegmentKiller uses this information to clean the version, interval and dataSource directories
   // on segment deletion if segment being deleted was the only segment
-  static String getDefaultStorageDir(DataSegment segment)
+  static String getDefaultStorageDir(DataSegment segment, boolean useUniquePath)
   {
     return JOINER.join(
         segment.getDataSource(),
         StringUtils.format("%s_%s", segment.getInterval().getStart(), segment.getInterval().getEnd()),
         segment.getVersion(),
-        segment.getShardSpec().getPartitionNum()
+        segment.getShardSpec().getPartitionNum(),
+        useUniquePath ? generateUniquePath() : null
     );
   }
+
+  static String generateUniquePath()
+  {
+    // There's a pretty low chance tasks publishing the same (dataSource, interval, version, shard) will collide so save
```

Comment:
Well, for there to be a collision, it has to happen as part of one of the failure scenarios, meaning that old shards from jobs run a long time ago aren't going to be involved in any collisions, even if there are millions of them. The scenarios would be things like task replicas both pushing to S3 at the same time, or a Kafka task failing after pushing or publishing and the replacement task pushing the same segment IDs, and in those cases, you'd have to have configured something really wrong to be pushing 100k segments from an indexing task.

But if you think it's still safer, I'm okay with using the full UUID; just trying to do my part to keep people from hitting their `PATH_MAX` or whatever on their file system.

Reply:
I'm thinking of the situation where you have two replicas and they will always push to S3 around the same time (which seems likely). Even in a non-failure scenario there will still be two pushes going on at once. In that case it's likely that some pair of replicas will collide with itself.

Reply:
> I'm thinking of the situation where you have two replicas and they will always push to S3 around the same time (which seems likely).

Ah, I missed this case. If so, what do you think about receiving a suffix as a parameter and making the caller to be responsible for passing a unique suffix as [commented above](https://github.com/druid-io/druid/pull/5692#discussion_r183902889)?

Reply:
I'd generally prefer random instead of something like an attempt id. It's because with attempt ids, you have to worry about not assigning two tasks the same attempt id by accident, and we don't have a concept like this right now. With big random numbers you know for sure that you are ok, without needing to think about it too hard.

Reply:
Okay sounds good to me